File: backend/src/models/FavoriteTransaction.js
const mongoose = require('mongoose');

const favoriteTransactionSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  name: {
    type: String,
    required: true,
    trim: true,
    maxlength: [100, 'Name cannot exceed 100 characters']
  },
  categoryKey: {
    type: String,
    required: true,
    uppercase: true,
    trim: true
  },
  itemId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'POSItem',
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    min: [1, 'Quantity must be at least 1'],
    default: 1
  },
  totalAmount: {
    type: mongoose.Schema.Types.Decimal128,
    required: true,
    min: [0, 'Amount must be positive'],
    set: val => parseFloat(val).toFixed(2),
    get: val => parseFloat(val.toString())
  }
}, {
  timestamps: true,
  toJSON: { getters: true },
  toObject: { getters: true }
});

favoriteTransactionSchema.index({ userId: 1, createdAt: -1 });
favoriteTransactionSchema.index({ userId: 1, categoryKey: 1 });

module.exports = mongoose.model('FavoriteTransaction', favoriteTransactionSchema);

File: mobile-app/nfc_app/lib/features/pos/domain/favorite_transaction.dart
import 'pos_item.dart';

class FavoriteTransaction {
  const FavoriteTransaction({
    required this.id,
    required this.userId,
    required this.name,
    required this.categoryKey,
    required this.itemId,
    required this.quantity,
    required this.totalAmount,
    this.itemDetails,
    this.createdAt,
  });

  final String id;
  final String userId;
  final String name;
  final String categoryKey;
  final String itemId;
  final int quantity;
  final double totalAmount;
  final POSItem? itemDetails;
  final DateTime? createdAt;

  factory FavoriteTransaction.fromJson(Map<String, dynamic> json) {
    return FavoriteTransaction(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      categoryKey: json['categoryKey']?.toString() ?? '',
      itemId: json['itemId']?.toString() ?? '',
      quantity: (json['quantity'] as num?)?.toInt() ?? 1,
      totalAmount: (json['totalAmount'] as num?)?.toDouble() ?? 0,
      itemDetails: json['itemId'] is Map<String, dynamic> 
          ? POSItem.fromJson(json['itemId'] as Map<String, dynamic>)
          : null,
      createdAt: json['createdAt'] != null 
          ? DateTime.tryParse(json['createdAt'].toString()) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'categoryKey': categoryKey,
      'itemId': itemId,
      'quantity': quantity,
      'totalAmount': totalAmount,
    };
  }
}
